### PR TITLE
[LOLIPOP-INTELLIGENCE] iP

### PR DESCRIPTION
**Pull request by LOLIPOP-INTELLIGENCE**
# 🍎 Appleaster: Task Manager Update

## Changes:
1. Enhanced task creation
2. Improved deadline tracking

Practise your Java:
```java
    public static void main(String[] args) {
        new Appleaster("data/tasks.txt").run();
    }
```

- [x] Complete tests
- [ ] Update docs

> Simplify your life with Appleaster!

**Key feature**: *Natural language* processing for task input.

~~Minor updates~~ Major improvements!

### CODE REVIEW

File: src/main/java/com/appleaster/parser/Parser.java [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+     * @param input The user input string to parse.
+     * @return A Command object representing the parsed input.
+     * @throws AppleasterException If the input is invalid or cannot be parsed.
+     */
+    public static Command parseCommand(String input) throws AppleasterException {
+        String[] parts = input.split("\\s+", 2);
+        String commandWord = parts[0].toLowerCase();
+        String arguments = parts.length > 1 ? parts[1] : "";
+
+    switch (commandWord) {
```
Comment (Edsel-Tan) [inline src/main/java/com/appleaster/parser/Parser.java:34]: Wrong indentation level?

File: src/main/java/com/appleaster/parser/Parser.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+      case "list":
+        return new Command(CommandType.LIST);
+      case "mark":
```
Comment (Edsel-Tan) [inline src/main/java/com/appleaster/parser/Parser.java:37]: Is the fallthrough intentional? Should leave a comment

File: src/main/java/com/appleaster/main/Appleaster.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+     * Executes a command based on the parsed user input.
+     *
+     * @param c The Command object representing the user's input.
+     * @return true if the command is to exit the application, false otherwise.
+     * @throws AppleasterException If there's an error executing the command.
+     */
+    private boolean executeCommand(Command c) {
+        try {
+            switch (c.getType()) {
+                case LIST:
```
Comment (Edsel-Tan) [inline src/main/java/com/appleaster/main/Appleaster.java:38]: wrong indentation level

**Review by Edsel-Tan — COMMENTED** [review]
Code is super nice and clean! 👍

File: data/tasks.txt [hunk 4]
```diff
@@ -0,0 +1,2 @@
+T | 1 | call mom
```
Comment (blanklogic) [inline data/tasks.txt]: As data/tasks.txt is going to be constantly changing, might be good to not track it. Consider adding it to gitignore as well as removing this cached version to prevent any further tracking by git.

File: src/main/java/com/appleaster/command/CommandType.java [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package com.appleaster.command;
+public enum CommandType {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/command/CommandType.java:2]: Nice utilisation of Enums to simplify the finite Command Types.

File: src/main/java/com/appleaster/command/Command.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
+package com.appleaster.command;
+import java.time.LocalDate;
+
+import com.appleaster.task.Task;
+
+public class Command {
+    private final CommandType type;
+    private Task task;
```
Comment (blanklogic) [inline src/main/java/com/appleaster/command/Command.java:12]: The following fields could be effectively final as well, I noticed the rest of your code in this class do not further modify them after being initialised by the use of overloaded constructors

File: src/main/java/com/appleaster/command/Command.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+    public Command(CommandType type, LocalDate date) {
+        this.type = type;
+        this.date = date;
+    }
+
+    public CommandType getType() {
+        return type;
+    }
+
+    public Task getTask() {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/command/Command.java:88]: # Potential NullPointerException
- Since your overloaded constructors do not initialise the values of the other fields, only specific ones.
- Calling upon getTask() on a Command object constructed with public Command(CommandType type) might throw a NullPointerException.
- Consider including default values for the other fields if not in use to prevent this.

File: src/main/java/com/appleaster/exception/AppleasterException.java [hunk 8]
```diff
@@ -0,0 +1,6 @@
+package com.appleaster.exception;
+public class AppleasterException extends Exception {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/exception/AppleasterException.java:2]: Perhaps extending from RuntimeException might be more appropriate as AppleasterException is thrown during run time as opposed to during compile time.

File: src/main/java/com/appleaster/main/Appleaster.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+import com.appleaster.ui.Ui;
+
+/**
+ * The main class for the Appleaster application.
+ * This class initializes the application and handles the main interaction loop.
+ */
+
+public class Appleaster {
+  private final Storage storage;
+  private TaskList tasks;
```
Comment (blanklogic) [inline src/main/java/com/appleaster/main/Appleaster.java]: Can consider being made final as TaskList reference is not modified after being initialised by constructor

File: src/main/java/com/appleaster/main/Appleaster.java (hunk 3 above)
Comment (blanklogic) [inline src/main/java/com/appleaster/main/Appleaster.java:38]: The use of switch case code blocks do not require indentations according to SE-EDU Java Coding Standard, refer to the example below.

```java
switch (c.getType()) {
case LIST:
    task.listTask();
    break;
case MARK:
case UNMARK:
    /* Code omitted ... */
```

File: src/main/java/com/appleaster/parser/Parser.java [hunk 10]
```diff
@@ -0,0 +1,191 @@
...
+import com.appleaster.task.Todo;
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeParseException;
+
+/**
+ * Parses user input into Command objects for the Appleaster application.
+ */
+public class Parser {
+  private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ISO_LOCAL_DATE;
```
Comment (blanklogic) [inline src/main/java/com/appleaster/parser/Parser.java:18]: Wrong indentation level. Should be an indentation of 4 spaces.

File: src/main/java/com/appleaster/storage/Storage.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+        } catch (IOException e) {
+            throw new AppleasterException("Error saving tasks: " + e.getMessage());
+        }
+    }
+
+    private void createDataDirectoryIfNotExists() throws IOException {
+        Files.createDirectories(filePath.getParent());
+    }
+
+    private String formatTask(Task task) {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/storage/Storage.java:95]: Might be good to leave the String representation formatting to the classes themselves such as Todo.java, Deadline.java, Event.java. Furthermore, notice how there's a lot of overlaps in the code that have been repeated. Perhaps adding the common formatting to the abstract class Task.java and Overriding the String representation in it's children could make it easier to locate and make changes in the future.

tldr; OOP concept of abstraction barrier. Ask the task to return the String instead of asking information from the task to create your own String.

File: src/main/java/com/appleaster/task/Task.java [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+    public boolean isCompleted() {
+        return this.status == TaskStatus.DONE;
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    @Override
+    public abstract String toString();
```
Comment (blanklogic) [inline src/main/java/com/appleaster/task/Task.java:70]: As mentioned in another comment, might be good to initialise your toString() in Task.java to contain the key information that is utilised by all the following children Todo.java, Event.java, Deadline.java and Overriding them as necessary to add additional information to be more specific to the child.

File: src/main/java/com/appleaster/task/Task.java [hunk 13]
```diff
@@ -0,0 +1,31 @@
+package com.appleaster.task;
+public abstract class Task {
+    protected String description;
+    protected TaskStatus status;
```
Comment (blanklogic) [inline src/main/java/com/appleaster/task/Task.java:12]: A boolean of isCompleted should suffice instead of polluting the heap with more objects than needed.

File: src/main/java/com/appleaster/task/Deadline.java [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+        super(description, TaskType.DEADLINE);
+        this.by = LocalDateTime.parse(by, INPUT_FORMAT);
+    }
+
+    public LocalDateTime getBy() {
+        return by;
+    }
+
+    @Override
+    public String toString() {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/task/Deadline.java:45]: Utilise the OOP concept of inheritance by abstracting out the common information such as description to be returned by the superclass/parent Task.java instead before Overriding and adding more specific information to reduce the need to modify multiple files for a change.

File: src/main/java/com/appleaster/task/Event.java [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+    public LocalDateTime getFrom() {
+        return from;
+    }
+
+    public LocalDateTime getTo() {
+        return to;
+    }
+
+    @Override
+    public String toString() {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/task/Event.java:59]: Utilise the OOP concept of inheritance by abstracting out the common information such as description to be returned by the superclass/parent Task.java instead before Overriding and adding more specific information to reduce the need to modify multiple files for a change.

File: src/main/java/com/appleaster/task/TaskList.java [hunk 16]
```diff
@@ -0,0 +1,169 @@
+package com.appleaster.task;
+
+import com.appleaster.exception.AppleasterException;
```
Comment (blanklogic) [inline src/main/java/com/appleaster/task/TaskList.java:3]: Might have been a mistake of double importing but the import at Line 9 is appropriate. Line 3 should trigger a checkstyle error.

File: src/main/java/com/appleaster/task/TaskList.java [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+    public TaskList(List<Task> tasks) {
+        this.tasks = tasks;
+    }
+
+    /**
+     * Adds a new task to the list.
+     *
+     * @param task The task to be added.
+     */
+  public void addTask(Task task) {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/task/TaskList.java]: Wrong indentation level for all the code blocks after this line.

File: src/main/java/com/appleaster/task/TaskStatus.java [hunk 18]
```diff
@@ -0,0 +1,15 @@
+package com.appleaster.task;
+public enum TaskStatus {
```
Comment (blanklogic) [inline src/main/java/com/appleaster/task/TaskStatus.java:2]: Might be redundant to abstract TaskStatus out as an Enum when a isCompleted() and String formatting of X in Task.java should suffice.

**Review by blanklogic — COMMENTED** [review]
# Overview
There were quite a few violations of Java Coding Standards in the Code. But with the appropriate fixes suggested should be easy to rectify. Consider drawing out a UML Diagram to better make sense of the code structure for OOP so as to optimise the amount of Objects being created as well as to incorporate Software Engineering Practices such as Do not Repeat Yourself (DRY) and SOLID.

# Conclusion
Overall, after some fixes LGTM.
